File: tools/ghidra_scripts/FindProtection.java
// Decompile the copy-protection routine. The protection strings live in DGROUP;
// the code references them by DGROUP-relative offset (recovered from the unpack:
// DGROUP seg 0x204b, load base 0x1010). Scan instructions for those scalar
// offsets and decompile the containing functions.
// @category Bumpy
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.scalar.Scalar;
import java.util.LinkedHashMap;
import java.util.Map;

public class FindProtection extends GhidraScript {
    @Override
    public void run() throws Exception {
        // DGROUP-relative offsets of the protection strings (from find_string_refs.py).
        long[] vals = {0x1331, 0x0606, 0x12e7, 0x12f5, 0x1309, 0x1318};
        String[] lbls = {"Enter the platform number", "INSERT THE OTHER DISK",
                         "YOUR PASSWORD", "ENTER YOUR PASSWORD", "PASSWORD OK", "PASSWORD ERROR"};
        Map<Long, String> want = new LinkedHashMap<>();
        for (int i = 0; i < vals.length; i++) want.put(vals[i], lbls[i]);

        FunctionManager fm = currentProgram.getFunctionManager();
        Map<Address, Function> funcs = new LinkedHashMap<>();

        InstructionIterator it = currentProgram.getListing().getInstructions(true);
        while (it.hasNext()) {
            Instruction ins = it.next();
            for (int op = 0; op < ins.getNumOperands(); op++) {
                Scalar sc = ins.getScalar(op);
                if (sc == null) continue;
                long v = sc.getUnsignedValue() & 0xFFFFL;
                if (want.containsKey(v)) {
                    Function f = fm.getFunctionContaining(ins.getAddress());
                    println(String.format("REF %s : %-22s -> %#x (%s) func=%s",
                            ins.getAddress(), ins.toString(), v, want.get(v),
                            f == null ? "(none)" : f.getName()));
                    if (f != null) funcs.put(f.getEntryPoint(), f);
                }
            }
        }

        DecompInterface decomp = new DecompInterface();
        decomp.openProgram(currentProgram);
        for (Map.Entry<Address, Function> e : funcs.entrySet()) {
            Function f = e.getValue();
            println("");
            println("===== DECOMP " + f.getName() + " @ " + f.getEntryPoint() + " =====");
            DecompileResults res = decomp.decompileFunction(f, 120, monitor);
            if (res != null && res.getDecompiledFunction() != null)
                println(res.getDecompiledFunction().getC());
            else
                println("   <decompilation failed>");
        }
    }
}

File: tools/ghidra_scripts/FindProtection.py
# Find the copy-protection routine(s): locate protection strings, their code
# xrefs, and decompile the referencing functions.
# @category Bumpy
# Jython 2.7 compatible (no f-strings).
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

prog = currentProgram
mem = prog.getMemory()
refmgr = prog.getReferenceManager()
fm = prog.getFunctionManager()
monitor = ConsoleTaskMonitor()

targets = ["ENTER YOUR PASSWORD", "PASSWORD ERROR", "PASSWORD OK",
           "Enter the platform", "INSERT THE OTHER", "YOUR PASSWORD"]


def find_all(s):
    res = []
    b = bytearray(s.encode("latin1"))
    a = prog.getMinAddress()
    while True:
        hit = mem.findBytes(a, b, None, True, monitor)
        if hit is None:
            break
        res.append(hit)
        a = hit.add(1)
    return res


decomp = DecompInterface()
decomp.openProgram(prog)

seen = set()
func_for_string = {}

for s in targets:
    addrs = find_all(s)
    for addr in addrs:
        print("STRING %r @ %s" % (s, addr))
        refs = list(refmgr.getReferencesTo(addr))
        if not refs:
            print("   (no direct xrefs found)")
        for r in refs:
            fa = r.getFromAddress()
            f = fm.getFunctionContaining(fa)
            fname = f.getName() if f is not None else "(none)"
            print("   xref from %s  reftype=%s  func=%s" % (fa, r.getReferenceType(), fname))
            if f is not None:
                func_for_string.setdefault(f.getEntryPoint(), f)

print("")
print("==================== DECOMPILED REFERENCING FUNCTIONS ====================")
for ep, f in func_for_string.items():
    if ep in seen:
        continue
    seen.add(ep)
    res = decomp.decompileFunction(f, 90, monitor)
    print("")
    print("===== %s @ %s =====" % (f.getName(), ep))
    if res is not None and res.getDecompiledFunction() is not None:
        print(res.getDecompiledFunction().getC())
    else:
        print("   <decompilation failed>")
